packages/devtools_app/lib/src/screens/vm_developer/object_inspector/vm_ic_data_display.dart

```diff
@@ -0,0 +1,129 @@
+// Copyright 2023 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/material.dart';
+import 'package:vm_service/vm_service.dart';
+
+import '../../../shared/common_widgets.dart';
+import '../../../shared/globals.dart';
+import '../vm_developer_common_widgets.dart';
+import '../vm_service_private_extensions.dart';
+import 'object_inspector_view_controller.dart';
+import 'vm_object_model.dart';
+
+// TODO(mtaylee): Finish implementation of [ICDataArrayWidget] and add it to
+// the [VmFuncDisplay].
+
+/// A widget for the object inspector historyViewport displaying information
+/// related to function (Func type) objects in the Dart VM.
+class VmICDataDisplay extends StatefulWidget {
+  const VmICDataDisplay({
+    required this.controller,
+    required this.icData,
+  });
+
+  final ObjectInspectorViewController controller;
+  final ICDataObject icData;
+
+  @override
+  State<VmICDataDisplay> createState() => _VmICDataDisplayState();
+}
+
+class _VmICDataDisplayState extends State<VmICDataDisplay> {
+  final argumentsDescriptor = <ObjRef?>[];
+  final entries = <ObjRef?>[];
+
+  late Future<void> _initialized;
+
+  @override
+  void initState() {
+    super.initState();
+    _initialize();
+  }
+
+  @override
+  void didUpdateWidget(VmICDataDisplay oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.icData == oldWidget.icData) {
+      return;
+    }
+    _initialize();
+  }
+
+  void _initialize() async {
+    argumentsDescriptor.clear();
+    entries.clear();
+
+    void populateLists(Instance argDescriptor, Instance entryList) {
+      argumentsDescriptor.addAll(argDescriptor.elements!.cast<ObjRef?>());
+      entries.addAll(entryList.elements!.cast<ObjRef?>());
+    }
+
+    final icData = widget.icData.obj;
+    if (icData.argumentsDescriptor is Instance && icData.entries is Instance) {
+      populateLists(
+        icData.argumentsDescriptor as Instance,
+        icData.entries as Instance,
```

Comment:
nit: if we store `icData.argumentsDescriptor` and `icData.entries` as variables, these will be type promoted after the if and we can remove these casts